ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+                then
+                  (* Optimization: update [v.lower] based on [mlower u].*)
+                  let mu_lower = mlower obj mu in
+                  if not (C.le obj mu_lower v.lower)
+                  then update_lower ?log obj v mu_lower);
+                r)
+       in
+       if C.le obj v.upper v.lower then set_vlower ?log v [];
+       r)
+
+  and submode_mvc :
+        'a 'r.
+        ?log:change list ref ->
+        'a C.obj ->
+        ('a, allowed * 'r) morphvar ->
+        'a ->
+        (unit, 'a) Result.t =
+   fun ?log obj (Amorphvar (v, f) as mv) a ->
+    (* See [submode_cmv] for why we need the following seemingly redundant
+       lines. *)
+    if C.le obj (mupper obj mv) a
+    then Ok ()
+    else if not (C.le obj (mlower obj mv) a)
+    then Error (mlower obj mv)
+    else
+      let a = C.meet obj a (mupper obj mv) in
+      let f' = C.right_adjoint obj f in
+      let src = C.src obj f in
+      let a' = C.apply src f' a in
+      (* cannot assert [Option.is_none] because [mlower] is imprecise. *)
+      Result.map_error (C.apply obj f) (submode_vc ?log src v a')
+
+  (** Calculate the precise lower bound *)
+  let zap_to_floor_try (type a) (obj : a C.obj) (v : a var) =
+    let rec loop lower =
+      let log = ref [] in
+      let r = submode_vc ~log obj v lower in
+      match r with
+      | Ok () -> log, lower
+      | Error a ->
+        undo_changes !log;
+        loop (C.join obj a lower)
+    in
+    loop v.lower
+
+  let constrain_mlower_try dst (Amorphvar (v, f)) =
+    let src = C.src dst f in
+    let log, lower = zap_to_floor_try src v in
+    log, C.apply dst f lower
+
+  let eq_morphvar :
+      type a l0 r0 l1 r1.
+      a C.obj -> (a, l0 * r0) morphvar -> (a, l1 * r1) morphvar -> bool =
+   fun dst (Amorphvar (v0, f0)) (Amorphvar (v1, f1)) ->
+    let obj0 = C.src dst f0 in
+    let obj1 = C.src dst f1 in
+    match C.eq_obj obj0 obj1 with
+    | None -> false
+    | Some Refl ->
+      v0 == v1
+      && C.disallow_left (C.disallow_right f0)
+         = C.disallow_left (C.disallow_right f1)
+
+  let submode_mvmv (type a) ?log (dst : a C.obj) (Amorphvar (v, f) as mv)
+      (Amorphvar (u, g) as mu) =
+    if C.le dst (mupper dst mv) (mlower dst mu)
+    then Ok ()
+    else if eq_morphvar dst mv mu
+    then Ok ()
+    else
+      (* we have f v <= g u which gives g' (f v) <= u. On the other hand, we
+         also have v <= f' (g u) *)
+      match submode_mvc ?log dst mv (mupper dst mu) with
+      | Error a -> Error (a, mupper dst mu)
+      | Ok () -> (
+        match submode_cmv ?log dst (mlower dst mv) mu with
+        | Error a -> Error (mlower dst mv, a)
+        | Ok () ->
+          let g' = C.left_adjoint dst g in
+          let src = C.src dst g in
+          let g'f = C.compose src g' (C.disallow_right f) in
+          let x = Amorphvar (v, g'f) in
+          if not (List.exists (fun mv -> eq_morphvar src mv x) u.vlower)
+          then set_vlower ?log u (x :: u.vlower);
```

Comment:
Is this a good time to take the join of all the lower bounds to put into `u.lower`?

Reply:
It wouldn't hurt to do so, but it would mean to traverse the whole sub-tree of `u` and slower. (Note that above, we already traversed the sub-tree of `v`, but not that of `u`).

I think the point is "only update u.lower when we already have to traverse the subtree of u anyway".